### PR TITLE
drivers: mtd_spi_nor MTD interface driver for SPI NOR flash chips

### PR DESCRIPTION
**Pull request by vincent-d**
This is based on #5668 with some minor improvements with erase and this has been ported to new spi interface.

### CODE REVIEW

**Comment by gebart** [comment]
There's an accidental autosave file included in this PR

**Comment by vincent-d** [comment]
Fixed

File: boards/mulle/board.c [hunk 1]
```diff
@@ -238,3 +262,15 @@ static void increase_boot_count(void)
     ++rec.u32;
     mulle_nvram->write(mulle_nvram, &rec.u8[0], MULLE_NVRAM_BOOT_COUNT, sizeof(rec.u32));
 }
+
+int mulle_nor_init(void)
```
Comment (vincent-d) [inline boards/mulle/board.c]: Done

**Comment by gebart** [comment]
Needs rebase

**Comment by vincent-d** [comment]
rebased

File: drivers/mtd_spi_nor/mtd_spi_nor.c [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+    }
+    else if ((dev->flag & SPI_NOR_F_SECT_4K) && size == 4096) {
+        /* 4 KiO sectors can be erased with sector erase command */
+        mtd_spi_cmd_addr_write(dev, dev->opcode->sector_erase, addr_be, NULL, 0);
+    }
+    else if ((dev->flag & SPI_NOR_F_SECT_32K) && size == 32768) {
+        /* 4 KiO sectors can be erased with sector erase command */
+        mtd_spi_cmd_addr_write(dev, dev->opcode->block_erase_32k, addr_be, NULL, 0);
+    }
+    else {
```
Comment (gebart) [inline drivers/mtd_spi_nor/mtd_spi_nor.c]: `size` is not checked here. This causes the unit test for erase of invalid size to fail.

File: drivers/mtd_spi_nor/mtd_spi_nor.c [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+    mtd_spi_nor_t *dev = (mtd_spi_nor_t *)mtd;
+    uint32_t total_size = mtd->page_size * mtd->pages_per_sector * mtd->sector_count;
+
+    if (dev->sec_addr_mask &&
+        ((addr & ~dev->sec_addr_mask) != 0)) {
+        /* This is not a requirement in hardware, but it helps in catching
+         * software bugs (the erase-all-your-files kind) */
+        DEBUG("addr = %" PRIx32 " ~dev->erase_addr_mask = %" PRIx32 "", addr, ~dev->sec_addr_mask);
+        DEBUG("mtd_spi_nor_erase: ERR: erase addr not aligned on %" PRIu32 " byte boundary.\n",
+            sector_size);
```
Comment (gebart) [inline drivers/mtd_spi_nor/mtd_spi_nor.c]:
```
riot/drivers/mtd_spi_nor/mtd_spi_nor.c:447:13: error: ‘sector_size’ undeclared (first use in this function)
             sector_size);
             ^
```

File: drivers/mtd_spi_nor/mtd_spi_nor.c (hunk 2 above)
Comment (vincent-d) [inline drivers/mtd_spi_nor/mtd_spi_nor.c]: Addressed

File: drivers/mtd_spi_nor/mtd_spi_nor.c (hunk 3 above)
Comment (vincent-d) [inline drivers/mtd_spi_nor/mtd_spi_nor.c]: Addressed

File: drivers/mtd_spi_nor/mtd_spi_nor.c [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+        }
+        TRACE("\n");
+    }
+    /* Acquire exclusive access to the bus. */
+    spi_acquire(dev->spi, dev->cs, dev->mode, dev->clk);
+
+    do {
+        /* Send opcode followed by address */
+        spi_transfer_byte(dev->spi, dev->cs, true, opcode);
+        spi_transfer_bytes(dev->spi, dev->cs, true, (char *)addr_buf, NULL, dev->addr_width);
```
Comment (gebart) [inline drivers/mtd_spi_nor/mtd_spi_nor.c]: This currently keeps the CS pin asserted on erase commands (count = 0), causes unit test failures on Mulle
Fixed:
```c
        bool cont = (count > 0); /* only keep CS asserted when there is data that follows */
        spi_transfer_bytes(dev->spi, dev->cs, cont, (char *)addr_buf, NULL, dev->addr_width);
        /* Write data */
        if (cont) {
```

Comment (vincent-d) [inline drivers/mtd_spi_nor/mtd_spi_nor.c]: Fixed, thanks

**Comment by gebart** [comment]
```
mtd_tests.test_mtd_erase (tests/unittests/tests-mtd/tests-mtd.c 152) exp 0 was -139
```
```c
    /* Erase 2nd - 3rd sector */
    ret = mtd_erase(dev, dev->pages_per_sector * dev->page_size,
                    dev->pages_per_sector * dev->page_size * 2);
    TEST_ASSERT_EQUAL_INT(0, ret);
```
erase only supports a single page in the current implementation

```
mtd_tests.test_mtd_write_read (tests/unittests/tests-mtd/tests-mtd.c 209) exp -139 was 9
```
```c
    /* out of bounds write (addr) */
    ret = mtd_write(dev, buf, dev->pages_per_sector * dev->page_size * dev->sector_count,
                    sizeof(buf));
    TEST_ASSERT_EQUAL_INT(-EOVERFLOW, ret);
```
not sure what happens here.

**Comment by gebart** [comment]
https://github.com/gebart/RIOT/commit/451af3d8e23a36e76dd00167562cacf5030247c4 has some configuration for the Mulle MTD flash

Depends on #6810

**Comment by vincent-d** [comment]
- Fixed unittests. It passes on our hardware.
 - check bounds white write and erase

**Comment by gebart** [comment]
OK to squash

**Comment by vincent-d** [comment]
Squashed

**Comment by gebart** [comment]
@vincent-d do you mind approving #6810? It's a very small changeset

**Comment by vincent-d** [comment]
\o/ Thanks @gebart

File: drivers/mtd_spi_nor/mtd_spi_nor_configs.c [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    .rdid            = 0x9f,
+    .wren            = 0x06,
+    .rdsr            = 0x05,
+    .wrsr            = 0x01,
+    .read            = 0x03,
+    .read_fast       = 0x0b,
+    .page_program    = 0x02,
+    .sector_erase    = 0x20,
+    .block_erase_32k = 0x52,
+    .block_erase     = 0xd8,
```
Comment (benpicco) [inline drivers/mtd_spi_nor/mtd_spi_nor_configs.c:39]: This is `64KB BLOCK ERASE`

File: drivers/mtd_spi_nor/mtd_spi_nor.c [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+        /* 32 KiO sectors can be erased with sector erase command */
+        mtd_spi_cmd_addr_write(dev, dev->opcode->block_erase_32k, addr_be, NULL, 0);
+    }
+    else if (size % sector_size != 0) {
+        return -EOVERFLOW;
+    }
+    else {
+        for (size_t i = 0; i < size / sector_size; i++) {
+            mtd_spi_cmd_addr_write(dev, dev->opcode->block_erase, addr_be, NULL, 0);
+            addr += sector_size;
```
Comment (benpicco) [inline drivers/mtd_spi_nor/mtd_spi_nor.c:481]: and sector_size != 64k

**Review by benpicco — COMMENTED** [review]
I think I just discovered an old bug that was never noticed in the last 4 years :fearful:

**Pull request merged**